### PR TITLE
Add option for character name colors in chat/speechbubbles

### PR DESCRIPTION
**Pull request by SlamBamActionman**
## About the PR

People complained about not having a wall of white. This adds back the wall of white, optionally.
Colored names are on by default, making the wall of white opt-in.
Does not change channel color behavior.

## Why / Balance

Discord thread with people asking for it: https://discord.com/channels/310555209753690112/1200030805095895050 

There are plans to change the name color palette to something that works better from a UI-standpoint than what it currently is at the time of this PR being made. However, for those that do not find that to be enough or would like to have even more constrast between local chat and comms, this option provides that.

## Technical details

Since the server includes the color in the message string when it gets sent to the client, the client has to remove it on its side to comply with client-side settings. `RemoveMessageTag` enables this, with the option to remove both normal tags and tags with parameters. 

## Media

image

- [x] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

## Breaking changes

**Changelog**

:cl:
- add: Added option to toggle character name colors.

### CODE REVIEW

**Review by deathride58 — COMMENTED** [review]
It'd probably be ideal to move the coloration itself from the server to the client rather than the route of stripping the tags on the client's side

**Review by metalgearsloth — CHANGES_REQUESTED** [review]
Do what bhijn said, manually parsing text is a bad idea.

**Comment by SlamBamActionman** [comment]
> It'd probably be ideal to move the coloration itself from the server to the client rather than the route of stripping the tags on the client's side

> Do what bhijn said, manually parsing text is a bad idea.

So I've tried implementing the color system clientside, but I'm bumping up against an issue I can't solve in a satisfying way. Since the message string is compiled and formatted on server-side, once it's sent to the client there is some data lost; for my usecase, it's which part of the string is the name. 

Unless there is a rework to the chat system I don't really see a way to do this clientside without parsing the string. Either I parse the string to remove the `color` tag, or I parse the string to learn where the name is (there is a handy `BubbleHeader` tag for that, but I still need to inject the `color` tag around it).

**Comment by SlamBamActionman** [comment]
So the whole name color feature is now implemented clientside with the latest commit; I don't think it's much better since it still has to parse the string, however with the new `[Name]` tag at least it's less volatile.

The reason I'm using a new tag instead of `[BubbleHeader]` is because the color feature uses the name's string length to determine the color, however due to the additional `[bold]` tag added for spoken lines, using `[BubbleHeader]` resulted in different colors for whispered and spoken lines.

This would and should be reworked when the new chat system drops, however until then I see no way to elegantly implement this.

**Comment by BellwetherLogic** [comment]
I can't speak to the coding issues but the colored text makes the chat inaccessible for me: my brain is too busy picking out all the different colors as they stream by to parse the text, to the point of distraction and nigh-unusability. I really hope this is implemented.

File: Content.Client/UserInterface/Systems/Chat/ChatUIController.cs [hunk 1]
```diff
@@ -845,6 +870,45 @@ public void Repopulate()
...
+    public string GetStringInsideTag(ChatMessage message, string tag)
+    {
+        var rawmsg = message.WrappedMessage;
+        var tagStart = rawmsg.IndexOf($"[{tag}]");
+        var tagEnd = rawmsg.IndexOf($"[/{tag}]");
+        if (tagStart < 0 || tagEnd < 0)
+            return "";
+        tagStart += tag.Length + 2;
+        return rawmsg.Substring(tagStart, tagEnd - tagStart);
+    }
```
Comment (deathride58) [inline Content.Client/UserInterface/Systems/Chat/ChatUIController.cs:893]: This is a copy-paste of `SpeechBubble.cs`'s `ExtractSpeechSubstring()`; it'd be acceptable to move it somewhere common that both can access to keep the code duplication down

Comment (SlamBamActionman) [reply Content.Client/UserInterface/Systems/Chat/ChatUIController.cs]: Done - they're now static methods in SharedChatSystem, matching the behavior of SanitizeAnnouncement.

**Comment by SlamBamActionman** [comment]
Oh no. This rebase did not go as expected. 

Shit.

**Pull request closed without merging**